google_guest_agent/events/events.go

```diff
@@ -0,0 +1,221 @@
+// behavior.
+type Config struct {
+	Watchers []string
+}
+
+// EventCb defines the callback interface between watchers and subscribers.
+type EventCb func(evType string, data interface{}, evData interface{})
+
+type eventSubscriber struct {
+	data interface{}
+	cb   EventCb
+}
+
+type eventNotif struct {
+	leaving bool
+	evType  string
+	data    interface{}
+}
+
+func init() {
+	err := initWatchers([]Watcher{
+		metadata.New(),
+	})
+	if err != nil {
+		logger.Errorf("Failed to initialize watchers: %+v", err)
+	}
+}
+
+// init initializes the known available event watchers.
+func initWatchers(watchers []Watcher) error {
+	for _, curr := range watchers {
+		// Ignore & log if we are accidentaly not properly setting the id.
+		if curr.ID() == "" {
+			return fmt.Errorf("invalid event watcher id, skipping.")
+		}
+		availableWatchers[curr.ID()] = curr
+	}
+	return nil
+}
+
+// New allocates and initializes a events Manager based on provided cfg.
+func New(cfg *Config) (*Manager, error) {
+	res := &Manager{
+		watchers:    availableWatchers,
+		subscribers: make(map[string][]*eventSubscriber),
+	}
+
+	// Align manager's config based on consumers provided watchers If it is
+	// passing in wanted/expected watchers, otherwise use all available ones.
+	if cfg != nil && len(cfg.Watchers) > 0 {
+		res.watchers = make(map[string]Watcher)
+		for _, curr := range cfg.Watchers {
+			// Report back if we don't know the provided watcher id.
+			if _, found := availableWatchers[curr]; !found {
+				return nil, fmt.Errorf("invalid/unknown watcher id: %s", curr)
+			}
+			res.watchers[curr] = availableWatchers[curr]
+		}
+	}
+
+	return res, nil
+}
+
+// Subscribe registers an event consumer/subscriber callback to a given event type, data
+// is a context pointer provided by the caller to be passed down when calling cb when
+// a new event happens.
+func (mngr *Manager) Subscribe(evType string, data interface{}, cb EventCb) {
+	mngr.subscribers[evType] = append(mngr.subscribers[evType],
+		&eventSubscriber{
+			data: data,
+			cb:   cb,
+		},
+	)
+}
+
+type eventProcessing struct {
+	mutex       sync.Mutex
+	watchersMap map[string]bool
+}
+
+func (ep *eventProcessing) add(evType string) {
+	ep.mutex.Lock()
+	defer ep.mutex.Unlock()
+	ep.watchersMap[evType] = true
+}
+
+func (ep *eventProcessing) del(evType string) int {
+	ep.mutex.Lock()
+	defer ep.mutex.Unlock()
+	delete(ep.watchersMap, evType)
+	return len(ep.watchersMap)
+}
+
+// Run runs the event manager, it will block until all watchers have given up/failed.
+func (mngr *Manager) Run(ctx context.Context) {
+	var waitGroup sync.WaitGroup
+
+	notifChan := make(chan *eventNotif)
+	leaving := false
+
+	go func(ctx context.Context, channel chan<- *eventNotif, wg sync.WaitGroup, leaving *bool) {
+		wg.Add(1)
+		defer wg.Done()
+
+		select {
+		case <-ctx.Done():
+			*leaving = true
+			channel <- &eventNotif{
+				leaving: true,
+			}
+		}
+	}(ctx, notifChan, waitGroup, &leaving)
```

Comment:
do we need to pass `waitGroup` here, it will be already accessible inside go routine right?
Also, if we are passing it down to the function shouldn't it be the address (`&waitGroup`)?

Reply:
The intent was to make it explicit what is being used in each go routine (but obviously I failed to keep consistency).

Reply:
I changed the `go func()` of all instances to receive only channels where we want to force/guard the channel direction. I removed all the other arguments.